### ISSUE TITLE
reshape with non-contiguous input has wrong gradient on CUDA (breaking einsum)

### ISSUE THREAD

**Issue by speedcell4** [pytorch/pytorch#30303]
## 🐛 Bug

`einsum`, which takes **incontiguous** tensor as input, can not back-propagate correct gradient

## To Reproduce

```python
from argparse import ArgumentParser

import torch
from torch import Tensor


def check_jacobian(x: Tensor, weight: Tensor):
    print(f'{"contiguous" if x.is_contiguous() else "incontiguous"} input tensor')

    y1 = x @ weight
    y2 = torch.einsum('...x,xy->...y', [x, weight])
    assert torch.allclose(y1, y2)

    x.grad = None
    y1.backward(torch.eye(2, device=x.device))
    print(f'matmul.jacobian => {x.grad.view(-1)}')

    x.grad = None
    y2.backward(torch.eye(2, device=x.device))
    print(f'einsum.jacobian => {x.grad.view(-1)}')


if __name__ == '__main__':
    argument_parser = ArgumentParser()
    argument_parser.add_argument('--device', type=int, default=-1)
    args = argument_parser.parse_args()

    torch.cuda.set_device(args.device)
    torch.manual_seed(42)
    device = torch.device(f'cuda:{args.device}')

    _x = torch.rand(1, 2, device=device)
    x1 = _x.expand(2, -1).requires_grad_()
    x2 = _x.expand(2, -1).contiguous().requires_grad_()
    w = torch.rand(2, 2, device=device)

    check_jacobian(x1, w)
    check_jacobian(x2, w)
```

this code will output as below, but as you notice `einsum.jacobian` of `incontiguous input tensor` is different from all the others.

```
incontiguous input tensor
matmul.jacobian => tensor([0.9877, 0.5621, 0.1289, 0.5221], device='cuda:0')
einsum.jacobian => tensor([0.5583, 0.5421, 0.5583, 0.5421], device='cuda:0')
contiguous input tensor
matmul.jacobian => tensor([0.9877, 0.5621, 0.1289, 0.5221], device='cuda:0')
einsum.jacobian => tensor([0.9877, 0.5621, 0.1289, 0.5221], device='cuda:0')
```

## Expected behavior

```
incontiguous input tensor
matmul.jacobian => tensor([0.9877, 0.5621, 0.1289, 0.5221], device='cuda:0')
einsum.jacobian => tensor([0.9877, 0.5621, 0.1289, 0.5221], device='cuda:0')
contiguous input tensor
matmul.jacobian => tensor([0.9877, 0.5621, 0.1289, 0.5221], device='cuda:0')
einsum.jacobian => tensor([0.9877, 0.5621, 0.1289, 0.5221], device='cuda:0')
```

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.6 LTS
GCC version: (Homebrew gcc 5.5.0_4) 5.5.0
CMake version: version 3.15.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 418.87.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] numpydoc==0.8.0
[conda] pytorch-crf               0.7.0                    pypi_0    pypi
[conda] pytorch-pretrained-bert   0.6.2                    pypi_0    pypi
[conda] pytorch-transformers      1.0.0                    pypi_0    pypi
[conda] torch                     1.3.1                    pypi_0    pypi
[conda] torch-scatter             1.4.0                    pypi_0    pypi
[conda] torch-struct              0.3                      pypi_0    pypi
[conda] torchcrf                  1.0.4                    pypi_0    pypi
[conda] torchsnooper              0.3                      pypi_0    pypi
[conda] torchtext                 0.4.0                    pypi_0    pypi

cc @ezyang @gchanan @zou3519 @jerryzh168 @SsnL @albanD @gqchen

**Comment by ezyang** [comment]
cc @t-vi

**Comment by t-vi** [comment]
Strange, last I looked, einsum despatched to bmm. Does bmm work?

**Comment by t-vi** [comment]
Before we continue: @speedcell4 thank you for reporting with a minimal reproducing example, this helps tremendously to track these things down!

Based on the minimal example, we can also have this even more stripped down example:
```
import torch
torch.manual_seed(42)
device = torch.device('cuda:0')

_x = torch.rand(1, 2, device=device)
x1 = _x.expand(2, -1).requires_grad_()

print(torch.autograd.grad(x1.reshape(1, 2, 2), x1, torch.eye(2, device=x1.device)[None]))
print(torch.autograd.grad(x1.contiguous().reshape(1, 2, 2), x1, torch.eye(2, device=x1.device)[None]))
```
giving `[[0.5, 0.5],[0.5, 0.5]]` vs. the correct `[[1.0, 0], [0, 1.0]]`.

**Comment by t-vi** [comment]
`as_strided` isn't gpu specific, so this has the same bug:
```
import torch
torch.manual_seed(42)
_x = torch.rand(1, 2)
x1 = _x.expand(2, -1).requires_grad_()

print(torch.autograd.grad(x1.as_strided((1, 2, 2), (0, 0, 1)), x1, torch.eye(2)[None]))
print(torch.autograd.grad(x1.contiguous().as_strided((1, 2, 2), (0, 2, 1)), x1, torch.eye(2)[None]))
```

**Comment by t-vi** [comment]
@SsnL if you beat me to an idea to fix this...

**Comment by speedcell4** [comment]
@t-vi Thank you for quick responding, but this is not only on CUDA, but also on CPU

**Comment by t-vi** [comment]
Indeed. I should have a fix soonish, until it is committed you probably want contiguous leaf nodes.

**Comment by t-vi** [comment]
Hm. Upon reflection, I would need whether it's a view of a leaf. Sigh.

**Comment by SsnL** [comment]
I'd argue that this is not a bug. 

`as_strided` is a special op than all other ops in that its behavior depends on input geometry. So implicitly, we think of it as having `input.strides()` as an additional input. See [here](https://github.com/pytorch/pytorch/blob/7c4b9042ababa32dcd49e8d35114c81c53027431/tools/autograd/templates/Functions.cpp#L1093-L1214) for details.

Therefore, `x.contiguous().as_strided()` and `x.as_strided()` could be very different ops (in terms of gradient computation). And I would say that the different gradients in @t-vi 's script is expected:
```py
import torch
torch.manual_seed(42)
_x = torch.rand(1, 2)
x1 = _x.expand(2, -1).requires_grad_()

print(torch.autograd.grad(x1.as_strided((2, 2), (0, 1)), x1, torch.eye(2)))
print(torch.autograd.grad(x1.contiguous().as_strided((2, 2), (2, 1)), x1, torch.eye(2)))
```

Due to this complication, we should avoid using `as_strided` as a computation graph edge as much as we can. 

Here in the original script, `reshape` uses `as_strided` when the geometry is view-able. This has been fixed in https://github.com/pytorch/pytorch/pull/28901, which makes `reshape` use `view` instead. Now on a master build the original script gives the same identity gradient in both cases.

**Comment by t-vi** [comment]
So I'm at a point where I need input. Fundamentally, what happens is that I think that step 0 in the `as_strided_backward`:
https://github.com/pytorch/pytorch/blob/7c4b9042ababa32dcd49e8d35114c81c53027431/tools/autograd/templates/Functions.cpp#L1225-L1227
produces gradients that are - when considered by themselves - incorrect.
The reason we seem to get away with that is that we implicitly assume that stride-0-dimensions will have been derived from some call like `expand` that _will_ be backpropagated through further down the road (so it does the `sum_to_size` that is the backward of this `expand` ahead of time).
This breaks when we (unexpectedly?) use the gradient that `as_strided_backward` produces. The two cases I see are
- hooks (might not be as bad),
- leaves that have stride 0.
The difficulty (new compared to what I knew when I was optimistic) is that we might not directly have the leave, but a view of a leaf (e.g. chained as_strided calls - in `einsum`'s case permute + reshape).

With this diagnosis, I see three options:
- disallow leaves with stride 0 (easy, but will break someone's code somewhere),
- disable the optimization to sum gradients for stride 0 inputs (easy, but we would not have the optimization if it were not useful, so it is likely unattractive, see),
- try to replace the optimization with fusing edges in the autograd graph -- i.e. combine consecutive "as_strided"-edges. This would mean to manually write the autograd graph recording for as_strided and that we would hook the edge not to the output of the previous as_strided but to the input. This seems somewhat elaborate and fusion in the autograd graph is somewhat unconventional. It might cover the most important use cases of the optimization though. (@SSnL?).

Any of the options are with enough impact that I would value input on which route to take.

**Comment by SsnL** [comment]
It is not really about the optimization at step 0. If we remove that, the following steps would do the equivalent thing for stride 0 dims.

**Comment by t-vi** [comment]
So the bug happens with PyTorch version 1.3.1 and but as @SsnL explains, it will not happen on PyTorch master (because `einsum` uses permute + reshape but not as_strided directly). Then we should just retitle it back to einsum and close it as fixed in master.

**Comment by t-vi** [comment]
OK, so really, for anything using `reshape` (i.e. including `einsum`), this is fixed by #28901, so master/nightlies will have the originally reported bug fixed.
The more general `as_strided` may be not that fortunate, but my understanding is that that is not intended for wide-spread use anyway, so it isn't that urgent.
